wpilibj/src/main/java/edu/wpi/first/wpilibj/OpModeRobot.java

```diff
@@ -0,0 +1,664 @@
+   *
+   * @param cls class to add
+   * @throws IllegalArgumentException if class does not meet criteria
+   */
+  public void addAnnotatedOpMode(Class<? extends OpMode> cls) {
+    Autonomous auto = cls.getAnnotation(Autonomous.class);
+    Teleop teleop = cls.getAnnotation(Teleop.class);
+    TestOpMode test = cls.getAnnotation(TestOpMode.class);
+    if (auto == null && teleop == null && test == null) {
+      throw new IllegalArgumentException(
+          "must be annotated with Autonomous, Teleop, or TestOpMode");
+    }
+    addAnnotatedOpModeImpl(cls, auto, teleop, test);
+  }
+
+  private void addAnnotatedOpModeClass(String name) {
+    // trim ".class" from end
+    String className = name.replace('/', '.').substring(0, name.length() - 6);
+    System.out.println("adding " + className);
+    Class<?> cls;
+    try {
+      cls = Class.forName(className);
+    } catch (ClassNotFoundException e) {
+      return;
+    }
+    Autonomous auto = cls.getAnnotation(Autonomous.class);
+    Teleop teleop = cls.getAnnotation(Teleop.class);
+    TestOpMode test = cls.getAnnotation(TestOpMode.class);
+    if (auto == null && teleop == null && test == null) {
+      return;
+    }
+    try {
+      addAnnotatedOpModeImpl(cls, auto, teleop, test);
+    } catch (IllegalArgumentException e) {
+      reportAddOpModeError(cls, e.getMessage());
+    }
+  }
+
+  private void addAnnotatedOpModeClassesDir(File root, File dir, String packagePath) {
+    File[] files = dir.listFiles();
+    if (files == null) {
+      return;
+    }
+    for (File file : files) {
+      if (file.isDirectory()) {
+        addAnnotatedOpModeClassesDir(root, file, packagePath);
+      } else if (file.getName().endsWith(".class")) {
+        String relPath = root.toPath().relativize(file.toPath()).toString().replace('\\', '/');
+        addAnnotatedOpModeClass(packagePath + "." + relPath);
+      }
+    }
+  }
+
+  /**
+   * Scans for classes in the specified package and all nested packages that are annotated with
+   * {@link Autonomous}, {@link Teleop}, or {@link TestOpMode} and registers them. It's necessary to
+   * call publishOpModes() to make the added modes visible to the driver station.
+   *
+   * @param pkg package to scan
+   */
+  public void addAnnotatedOpModeClasses(Package pkg) {
+    String packageName = pkg.getName();
+    String packagePath = packageName.replace('.', '/');
+    ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+    System.out.println("Scanning " + packageName + " " + packagePath);
+
+    try {
+      Enumeration<URL> resources = classLoader.getResources(packagePath);
+      while (resources.hasMoreElements()) {
+        URL resource = resources.nextElement();
+        System.out.println("resource " + resource);
+        if ("jar".equals(resource.getProtocol())) {
+          // trim ".class" from end
+          String jarPath = resource.getPath().substring(5, resource.getPath().indexOf('!'));
```

Comment:
How is this trimming ".class" from the end? It's trimming 5 characters from the start and whatever comes after the exclamation mark.